dagshub/data_engine/model/query.py

```diff
@@ -54,20 +54,30 @@ class FieldFilterOperand(enum.Enum):
 
 
 class DatasourceQuery:
-    def __init__(self, column_or_query: Optional[Union[str, "DatasourceQuery"]] = None, as_of: Optional[int] = None):
+    def __init__(
+        self,
+        column_or_query: Optional[Union[str, "DatasourceQuery"]] = None,
+        field_as_of: Optional[int] = None,
+        query_as_of: Optional[int] = None,
+        selects: Optional[List] = None,
+    ):
```

Comment:
Seems to me like there's some type confusion here between top level query and the leaf / intermediate node query objects:
1. query_as_of should only belong on the top level, why pass it into all the internal nodes?
2. Same for selects
3. I'll have a related comment on the `__getitem__` of the Datasource

Reply:
Meaning, the top level query is probably best defined as its own class...
Or at most, an explicit subclass of DatasourceQuery with additional fields, rather than the current thing which is like a fragile implicit subclass

Reply:
I think I'm confused since the whole class hierarchy doesn't make sense to me, in my head (with intuition from my knowledge of Pandas), when doing `ds["bla"]` a Query should be returned, not a Datasource/set.
Something like `ds["bla"].all()` will only make sense if it returns an array of the "bla" column instead of the whole dataset dataframe...
I think that's the root of my confusion, and that might mean that my review here and/or the change that pushes query_as_of and selects into the Query class doesn't make much sense anyway

Reply:
That was how I got it to work. There are probably other ways to make it work, but this is the one I grew into organically.

Right now it's all one big Datasource class that the user is interacting with, and the DatasourceQuery is a "DTO" container for the query tree + anything that might be related to that.

The problem for me was that I didn't want to complicate things by having the interface be "multimodal", as in if you do x you get back a datasource, and if you do y, you get back a query. The complications from that would mainly be code duplication. Think this:
```python
ds.as_of(t1).all() # as_of is on the datasource
(ds["aaa"] == "bb").as_of(t1).all() # as_of in on the query
```
Same goes for `.all()` even. Basically now I'll have to duplicate every function  that you could do on a Datasource on a Query. Which is doable with a `__getattr__` magic I guess, but it feels a bit smelly